subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/resolver/SourcePathProvider.kt

```diff
@@ -73,5 +73,5 @@ object SourcePathProvider {
 
 internal
 fun subDirsOf(dir: File): Collection<File> =
-    if (dir.isDirectory) dir.listFiles().filter { it.isDirectory }
+    if (dir.isDirectory) dir.listFiles().filter { it.isDirectory }.sortedBy { it.name }
```

Comment:
❓ Would it be possible to sort only once at https://github.com/gradle/kotlin-dsl/blob/e035db702e25034569f117bad4fb0c7f87d1e71a/subprojects/tooling-builders/src/main/kotlin/org/gradle/kotlin/dsl/tooling/builders/KotlinBuildScriptModelBuilder.kt#L305 ?

Reply:
That would fix the order instability too. But it would break the `gradleSource()` then `classpathSources` then `accessorsClassPath.src` order. Moreover, when some source is available more than once in the source path, the IDE usually present the first one (and allows to pick choose others). Sorting the whole thing once would also break the assumed same-as-compile-classpath order, potentially inducing surprising behavior. I also think that keeping the overall order makes it easier to troubleshoot issues if need be.

`classpathSources` is "unsorted" anyway, because of https://github.com/gradle/gradle/issues/5507. Sorting buckets that need to be sorted once would be a good tradeoff, I'll amend the PR in this regard.